### ISSUE TITLE
Feature Request: Fuzzy Finder for code suggestion tool (eg. pressing '.' then [tab] and begin typing)

### ISSUE THREAD

**Issue by kadlugan** [jupyter/notebook#3083]
Hi All, 

This is is a request for a better fuzzy autocompletion tool for jupyter notebooks (ipython).  

It would be a great help to have a simpler yet better fuzzy finder that made it easier to isolate functions relevant to a particular class.  

Currently the autocomplete for jupyter feels more like a spelling checker (you have to know what you are looking for, be able to spell it, and then scroll heavily to find it ).

I am coding using deep learning libraries which pytorch, keras, tensorflow, theano, and these have pretty convoluted and similar and potentially confusing APIs.  Fuzzy finding in these libraries would be a huge help.

eg some examples in keras API
keras.layers

e.g. typing .1d might give you all the operations:
- AvfPool**1D**
- AveragePooling**1D**
- Conv**1D**

e.g. type .cv2d might give you all the operations:
- **C**on**v2D**
- De**C**on**v2D**
- *but not* Cropping2D

For inspiration [Fuzzy finder](https://github.com/amjith/fuzzyfinder) is a python library which does this, it appears.
![fuzzy finder example](https://raw.githubusercontent.com/amjith/fuzzyfinder/master/screenshots/pgcli-fuzzy.gif)

Is it possible to incorporate this code, or something similar, into an upcoming version of jupyter. It would be a great help!!

For other related discussions, I have searched the codebase and found these two issues:
[Better autocompletion for our Python APIs #135](https://github.com/googledatalab/datalab/issues/135)
[Fuzzy search completion list #504](https://github.com/jonathanslenders/python-prompt-toolkit/issues/504) and haven't found a solution for this.  

I'd love to hear if anyone has found any solutions to this, or people who might be able to add their upvotes, or other considerations.

Thanks

**Comment by kadlugan** [comment]
Ok, so I have isolated (i think) the potential code change to the [completer.js](https://github.com/jupyter/notebook/blob/df2fc69ef59c559e4788d571106a03dd1768e0b3/notebook/static/notebook/js/completer.js) class

Are there any js code wizzes who can add a fuzzy search to this package. Below are some GitHub libraries that I found in js.  Sorry, my js is pretty limited. What might take you minutes, would take me hours
https://github.com/bevacqua/fuzzysearch
https://github.com/krisk/Fuse
https://github.com/atom/fuzzy-finder

**Comment by takluyver** [comment]
The fuzzy search would need to be in the kernel side - the frontend Javascript can only process what the kernel sends back to it. For Python completions, that means it's part of IPython. @Carreau has been doing some work on the completion machinery, but I don't think it includes fuzzy matches.

**Comment by kadlugan** [comment]
Thanks @takluyver,  Great info.   

I see so let me know if I have this right. In Python completions, we get the whole completions list from the kernel, and we are just filtering that list in the Javascript front end with the prefix search and truncating the visible list suggestions

So my next question is - isn't a fuzzy search of the current classes methods just a different way of front-end filtering and truncating the visible list?   Do we need to modify the kernel at all?

**Comment by Carreau** [comment]
> I see so let me know if I have this right. In Python completions, we get the whole completions list from the kernel, and we are just filtering that list in the Javascript front end with the prefix search and truncating the visible list suggestions

Yes, that is correct. Actually not exactly prefixes (because of `%` in amgics, but that's implementation detail).



> So my next question is - isn't a fuzzy search of the current classes methods just a different way of front-end filtering and truncating the visible list? Do we need to modify the kernel at all?

Yes we have to modify kernel, because the kernel does not send 'all' completions, that would be too heavy. It sends only the matching one. e.g `numpy.his<tab>` will only send what starts with `his` to make it bearable on the network and frontend. If you then press `t`, we can filter further without extra requests. If you delete `s` we would have to fetch a new batch of potential completions.

Does that make sens ?

**Comment by takluyver** [comment]
> I see so let me know if I have this right. In Python completions, we get the whole completions list from the kernel, and we are just filtering that list in the Javascript front end with the prefix search and truncating the visible list suggestions

No, the kernel does the prefix search. The frontend displays all the completions it gets.

(I think @Carreau misunderstood what you were asking here, so I'm replying as well to make sure that it's clear.)

There's one extra detail which makes it more confusing: the frontend also finds some simple completions based on words you've used in the current notebook. It does this independently of the kernel. But the kernel, which knows about your code, is the more important source of completions.

**Comment by kadlugan** [comment]
Thanks @takluyver and @Carreau, 

Gotcha, I now understand the issue and purpose of the kernel and front end a lot better.  

So, fuzzy search is possible, BUT it requires kernel modification.  

@Carreau, so you are doing some work on the completion mechanics. Are you able to provide some insight into what you are currently working with regards to this and your thoughts on a fuzzy match approach? 

I'm guessing there are certain considerations that you have in the code base, but how difficult would it be to include a fuzzy feature.  I don't have much experience in kernel design, so forgive any ignorance I have?  What code base are you using in the kernel, is it python or something like c, c++  etc? 

As mentioned above, I found a few js and python libraries out there that provide fuzzy algorithms. 
https://github.com/bevacqua/fuzzysearch/blob/master/index.js . This one loops through each available method, and if each new character is in the tails of each of the current matched completions then it remains as part of the matched set.   (at least thats my initial perception)

@Carreau Im guessing this issue can be linked across to the kernel package. Can you suggest a place?

**Comment by takluyver** [comment]
> What code base are you using in the kernel, is it python or something like c, c++ etc?

It's Python. The code is split across two packages, [ipykernel](https://github.com/ipython/ipykernel) and [ipython](https://github.com/ipython/ipython).

The starting point for following the completions machinery is [this function](https://github.com/ipython/ipykernel/blob/cea3d93979b47014c3ec7a7f6ad203363098a7b0/ipykernel/ipkernel.py#L260). After that, it dives into `IPython.core.completer`.

**Comment by kadlugan** [comment]
Ok, thanks for that @takluyver i will have a look

**Comment by Carreau** [comment]
> so you are doing some work on the completion mechanics ...

Let's say that's "under my radar", there is no currently heavy active work. There are couple [PRs](https://github.com/ipython/ipython/pulls) about completions.

There is likely a lot of simplification that could be done now that we've drop Python 2; and I'd like to push most of what IPython does now into Jedi/Parso itself. 

Now that we are also using Prompt-toolkit in IPython there is a lot of work that could be done in making the completer asynchronous. 

One other possibility – which is harder, as it touches many packages is the following: 

When Code is running , we can't trigger completions in the notebook. Though because Jedi does support purely static analysis of code, we could lift this restriction and have less-exaustive, but still have completions in the notebook. 

Hope that give you some insights.

**Comment by swfsql** [comment]
just linking a newer issue about this: https://github.com/ipython/ipython/issues/12673
